crates/noirc_evaluator/src/ssa/acir_gen.rs

```diff
@@ -0,0 +1,720 @@
+//         }
+//     }
+
+//     if next_a.is_none() {
+//         *i2 += 1;
+//         return;
+//     }
+//     if next_b.is_none() {
+//         *i1 += 1;
+//         return;
+//     }
+//     if next_a.unwrap() < next_b.unwrap() {
+//         *i1 += 1;
+//         return;
+//     }
+//     *i2 += 1;
+// }
+
+// returns a + k*b
+pub fn add(a: &Arithmetic, k: FieldElement, b: &Arithmetic) -> Arithmetic {
+    let mut output = Arithmetic::default();
+
+    //linear combinations
+    let mut i1 = 0; //a
+    let mut i2 = 0; //b
+    while i1 < a.linear_combinations.len() && i2 < b.linear_combinations.len() {
+        if a.linear_combinations[i1].1 < b.linear_combinations[i2].1 {
+            output.linear_combinations.push(a.linear_combinations[i1]);
+            i1 += 1;
+        } else if a.linear_combinations[i1].1 > b.linear_combinations[i2].1 {
+            let coef = b.linear_combinations[i2].0 * k;
+            if coef != FieldElement::zero() {
+                output
+                    .linear_combinations
+                    .push((coef, b.linear_combinations[i2].1));
+            }
+            i2 += 1;
+        } else {
+            let coef = a.linear_combinations[i1].0 + b.linear_combinations[i2].0 * k;
+            if coef != FieldElement::zero() {
+                output
+                    .linear_combinations
+                    .push((coef, a.linear_combinations[i1].1));
+            }
+            i2 += 1;
+            i1 += 1;
+        }
+    }
+    while i1 < a.linear_combinations.len() {
+        output.linear_combinations.push(a.linear_combinations[i1]);
+        i1 += 1;
+    }
+    while i2 < b.linear_combinations.len() {
+        let coef = b.linear_combinations[i2].0 * k;
+        if coef != FieldElement::zero() {
+            output
+                .linear_combinations
+                .push((coef, b.linear_combinations[i2].1));
+        }
+        i2 += 1;
+    }
+
+    //mul terms
+
+    i1 = 0; //a
+    i2 = 0; //b
+    while i1 < a.mul_terms.len() && i2 < b.mul_terms.len() {
+        if (a.mul_terms[i1].1, a.mul_terms[i1].2) < (b.mul_terms[i2].1, b.mul_terms[i2].2) {
+            output.mul_terms.push(a.mul_terms[i1]);
+            i1 += 1;
+        } else if (a.mul_terms[i1].1, a.mul_terms[i1].2) > (b.mul_terms[i2].1, b.mul_terms[i2].2) {
+            let coef = b.mul_terms[i2].0 * k;
+            if coef != FieldElement::zero() {
+                output
+                    .mul_terms
+                    .push((coef, b.mul_terms[i2].1, b.mul_terms[i2].2));
+            }
+            i2 += 1;
+        } else {
+            let coef = a.mul_terms[i1].0 + b.mul_terms[i2].0 * k;
+            if coef != FieldElement::zero() {
+                output
+                    .mul_terms
+                    .push((coef, a.mul_terms[i1].1, a.mul_terms[i1].2));
+            }
+            i2 += 1;
+            i1 += 1;
+        }
+    }
+    while i1 < a.mul_terms.len() {
```

Comment:
I think my version is easier to reason with because it is the always the same pattern: all loops are while loop and use the same i1 and i2.

Reply:
Fair enough. The two sets of 3 loops you have look pretty similar though. I think you'd be able to factor them out if each mul_term and linear_combination had its own type in Arithmetic. E.g.
```rs
pub struct Arithmetic {
    pub mul_terms: Vec<MulTerm>,
    pub linear_combinations: Vec<LinearCombination>,
    pub q_c: FieldElement,
}
```
That way you could implement `Ord` for them and factor out this function into two calls of a helper, passing in which array you want I think. Hopefully it'd simplify other functions as well. It's a somewhat opinionated change and I've not investigated it myself in terms of how it'd affect the rest of the PR. I'll mark this as resolved to avoid blocking the PR over it